src/pointer-analysis/dereference.cpp

```diff
@@ -338,128 +347,83 @@ void dereferencet::dereference_addrof_expr(
+    {
+      // We can't reconstruct this. The base types are incompatible.
+      bad_base_type_failure(
+        guard, get_type_id(*expr->type), get_type_id(*base_of_steps_type));
+      return expr2tc();
+    }
+
     // Determine offset accumulated to this point (in bits)
-    expr2tc size_check_expr = expr;
-    wrap_in_scalar_step_list(size_check_expr, &scalar_step_list, guard);
-    expr2tc offset_to_scalar = compute_pointer_offset_bits(size_check_expr);
+    expr2tc offset_to_scalar = compute_pointer_offset_bits(base);
     simplify(offset_to_scalar);
 
     dereference2t &deref = to_dereference2t(expr);
     // first make sure there are no dereferences in there
     dereference_expr(deref.value, guard, dereferencet::READ);
 
-    const type2tc &to_type = scalar_step_list.back()->type;
+    const type2tc &to_type = base->type;
 
     expr2tc result =
       dereference(deref.value, to_type, guard, mode, offset_to_scalar);
     return result;
   }
 
-  if(is_index2t(expr) && is_pointer_type(to_index2t(expr).source_value))
-  {
-    index2t &index = to_index2t(expr);
-
-    // Determine offset accumulated to this point (computed in bytes)
-    expr2tc size_check_expr = expr;
-    wrap_in_scalar_step_list(size_check_expr, &scalar_step_list, guard);
-    expr2tc offset_to_scalar = compute_pointer_offset_bits(size_check_expr);
-    simplify(offset_to_scalar);
-
-    // first make sure there are no dereferences in there
-    dereference_expr(index.source_value, guard, dereferencet::READ);
-    dereference_expr(index.index, guard, dereferencet::READ);
-
-    add2tc tmp(index.source_value->type, index.source_value, index.index);
-
-    const type2tc &to_type = scalar_step_list.back()->type;
-
-    expr2tc result = dereference(tmp, to_type, guard, mode, offset_to_scalar);
-    return result;
-  }
-
   if(is_typecast2t(expr))
   {
     // Just blast straight through
     return dereference_expr_nonscalar(
-      to_typecast2t(expr).from, guard, mode, scalar_step_list);
+      to_typecast2t(expr).from, guard, mode, base);
   }
 
   if(is_member2t(expr))
-  {
-    scalar_step_list.push_front(expr);
-    expr2tc res = dereference_expr_nonscalar(
-      to_member2t(expr).source_value, guard, mode, scalar_step_list);
-    scalar_step_list.pop_front();
-    return res;
-  }
+    return dereference_expr_nonscalar(
+      to_member2t(expr).source_value, guard, mode, base);
 
   if(is_index2t(expr))
   {
-    dereference_expr(to_index2t(expr).index, guard, dereferencet::READ);
-    scalar_step_list.push_front(expr);
-    expr2tc res = dereference_expr_nonscalar(
-      to_index2t(expr).source_value, guard, mode, scalar_step_list);
-    scalar_step_list.pop_front();
+    index2t &index = to_index2t(expr);
+    dereference_expr(index.index, guard, dereferencet::READ);
+    expr2tc res =
+      dereference_expr_nonscalar(index.source_value, guard, mode, base);
     return res;
   }
 
   if(is_if2t(expr) && !is_scalar_type(expr))
   {
+    assert(0);
```

Comment:
Shall we keep this assertion here? I believe we are now outputting a message and then aborting.

Reply:
I tried but haven't succeeded to come up with an example that triggers this code path. Also the regression tests don't. It seems to be dead code. I'm ok with either removing the assertion in favour of a comment saying "dead: remove this branch", but I didn't want to test SV-COMP without it, yet. The assertion is compiled out for the SV-COMP benchmark runs, so in case it is used after all, at least the other (main) changes of this PR are not affected.

I could start another SV-COMP run without this branch as part of this PR, if you like.

Reply:
> I could start another SV-COMP run without this branch as part of this PR, if you like.

Can you please run another run over SV-COMP without this branch/assertion?

Reply:
Confirmed. The `if2t` case is also dead for SV-COMP: stats diff master/now: https://github.com/esbmc/esbmc/pull/734/files#diff-eaa56d75b0df4075b5b34397706a4e5cd062a446c7250304bb873d6068d00601